### ISSUE TITLE
Add API to access "extended file attributes" (xattr, EA)

### ISSUE THREAD

**Issue by heinrich-ulbricht** [dotnet/runtime#49604]
## Background and Motivation

The `File` API currently lacks support to create, read, update and delete "extended file attributes". What are extended file attributes? Citation from [the Wikipedia article](https://en.wikipedia.org/wiki/Extended_file_attributes):

> Extended file attributes are file system features that enable users to associate computer files with metadata not interpreted by the filesystem, whereas regular attributes have a purpose strictly defined by the filesystem (such as permissions or records of creation and modification times).

On *nix based systems they are called _xattr_, on Windows _E_ xtended _A_ ttributes (_EA_).

MacOS documentation: https://ss64.com/osx/xattr.html
Linux documentation: https://man7.org/linux/man-pages/man7/xattr.7.html
Windows documentation (?): https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-fscc/a82e9105-2405-4e37-b2c3-28c773902d85?redirectedfrom=MSDN (`$EA` and `$EA_INFORMATION`)
Hints to usage of EA on Windows: https://superuser.com/q/396692/93905

My use case specifically is storing an ID for files that is not visible for the user but links those files to external cloud sources those files were generated (and need to be update) from. Internet Explorer seems to store the "downloaded from Internet" information there, Antivirus vendors store scan information, Dropbox used those properties in the past as well.

Extended file attributes are a lightweight, cross-platform way of storing file metadata that cannot be tinkered with (at least not easily) by the user. Platform/file system support seems to be good. There are even recent developments like Linux NFS support: [User Xattr Support Finally Landing For NFS In Linux 5.9](https://www.phoronix.com/scan.php?page=news_item&px=Linux-5.9-NFS-Server-User-Xattr) 

## Proposed API

As extension to `System.IO.File`?
```
namespace System.IO
{
    public static class File
    {
        public static FileExtendedAttributes GetExtendedAttributes(string path);
    }
}
```

(Note: Like the existing _File.GetAttributes_ API)

(Note: It's also possible to set those attributes on directories.)

## Usage Examples
```
try
{
    // FileExtendedAttributes behaves like a Dictionary<string, string>
    FileExtendedAttributes extendedAttributes = File.GetExtendedAttributes("Program.cs");
    extendedAttributes.Add("key", "value");
    Debug.WriteLine(string.Join(", ", extendedAttributes.Keys)); // "key"
    Debug.WriteLine(string.Join(", ", extendedAttributes.Values)); // "value"
    extendedAttributes.Remove("key");
}
catch (NotSupportedException)
{
    // not supported by the underlying platform, file system or kernel
}

```

## Alternative Designs

The API design should reflect existing implementations from other languages and/or platforms. If this proposal is deemed worthy of being pursued further then we would have to look deeper into existing designs.

## Risks

Effort for a rarely-used feature.

Short-sighted design; searching for files based on those property values could be desirable. How could this be designed in contrast to existing designs?

**Comment by heinrich-ulbricht** [comment]
Area label would be _area-System.IO_ I suppose.

**Comment by carlossanlop** [comment]
We don't yet have a public type called `FileExtendedAttributes`. Can you please add it to your proposal? The design has to keep in mind that this is cross platform.

**Comment by heinrich-ulbricht** [comment]
This needs a deeper look into the platform-specific capabilities of the existing tools. Will hopefully have a look at it soon.

**Comment by heinrich-ulbricht** [comment]
Well Mr. MsftBot the label change was a bit too early...

**Comment by carlossanlop** [comment]
@heinrich-ulbricht have you had a chance to look into this more deeply?

@mklement0 [mentioned](https://github.com/dotnet/runtime/pull/52235#discussion_r626011438) the following today:

> _Note that macOS has open-ended extended file-system attributes, similar to NTFS alternate streams (use xattr -s -l /tmp to see an example on macOS),_

Another thing we need to keep in mind is that Windows has both extended attributes and reparse points. See [this document](https://docs.microsoft.com/en-us/windows/win32/fileio/reparse-points).

**Comment by heinrich-ulbricht** [comment]
Unfortunately not. It is relatively easy to propose something that would serve my specific use case. But I feel the need to be more thorough so that the proposal encompasses all platforms and respective platform features of xattr. Having found xattr only recently myself I'm a bit hesitant to propose anything to be honest ^^

**Comment by mklement0** [comment]
Thanks for starting this proposal, @heinrich-ulbricht.

Regarding the terminology on Windows: it seems to me that the equivalent of _extended attributes_ on Linux and macOS / BSD are _NTFS alternate data streams_; the table you link to uses attribute names _used by NTFS itself_ (whereas, as stated in the initial post, extended attributes are about user-defined payloads that have no special meaning to the file-system or the system).

If I understand the linked NTFS documentation correctly, the equivalent of a single extended attribute (a user-defined key-value pair containing arbitrary data) in the Unix world is a single alternate data stream on Windows (NTFS).

**Comment by heinrich-ulbricht** [comment]
@mklement0 I see, I misinterpreted the NTFS internal attribute name.

I worked with ADS in the (pre .NET Core) past, sounds interesting as an EA equivalent. Does this mean that .NET would have to provide a translation from an ADS API to EA file system features on Linux et al.? And that this does not yet exist? A quick search did not reveal anything other than old posts complaining that .NET access to ADS is not possible at all or about quirks like [PowerShell restrictions and/or bugs](https://blogs.msmvps.com/alunj/2020/06/24/revisiting-ntfs-alternate-data-streams/) (further down the page).

I also found a StackExchange question describing what I need: "[Is there a standardised way of adding custom metadata to image and other filetypes?](https://softwareengineering.stackexchange.com/questions/420111/is-there-a-standardised-way-of-adding-custom-metadata-to-image-and-other-filetyp)" 

What's your take on which direction this proposal might take?

**Comment by KalleOlaviNiemitalo** [comment]
The relevant Windows functions are [ZwQueryEaFile](https://docs.microsoft.com/windows-hardware/drivers/ddi/ntifs/nf-ntifs-zwqueryeafile) and [ZwSetEaFile](https://docs.microsoft.com/windows-hardware/drivers/ddi/ntifs/nf-ntifs-zwseteafile), but the documentation seems to cover kernel-mode calls only. There are also `FILE_READ_EA` and `FILE_WRITE_EA` in [File Access Rights Constants](https://docs.microsoft.com/windows/win32/fileio/file-access-rights-constants). These are not the same as named streams.

On Linux, I think .NET should treat the extended attribute namespace as part of the string, rather than automatically insert or delete a "user." prefix. That way, the API would be usable with all namespaces, including ones defined in the future. Such an API would make it harder to use extended attributes in a way that is portable across operating systems, but I don't know whether developers even want to do so; perhaps it is more important to read and write OS-specific extended attributes that are already in use.

**Comment by mklement0** [comment]
@heinrich-ulbricht, yes, as far as I'm aware there are no .NET APIs for ADS (NTFS alternate data streams) yet, though PowerShell's cmdlets do support them.

> What's your take on which direction this proposal might take?

I think a platform-neutral abstraction along the lines you've proposed would be beneficial, and seems possible at least in principle, given that what the platform-specific features share in the abstract is the ability to attach name-value pairs of arbitrary data to file-system items - but there are many details to be hashed out, and it'll be a challenge to find the right balance between focusing on shared, platform-neutral functionality vs. not preventing use of platform-specific functionality.

@KalleOlaviNiemitalo, I think we need to get clarity on the shared _concepts_ as opposed to _terminology_.

Even though in terms of _names_,  the Windows EAs (extended attributes) seem to be a closer fit, _conceptually_, ADS seem to be the equivalent of the Unix world extended attributes, i.e. arbitrary, user-defined pieces of data associated with file-system items that have no special meaning to either the file-system or the operating system.

Can you clarify how the Windows EAs fit into the picture here?

**Comment by iSazonov** [comment]
(If somebody want to try https://www.powershellgallery.com/packages/PSReflect-Functions/1.1/Content/Examples%5CGet-ExtendedAttribute.ps1)

**Comment by iSazonov** [comment]
[Here](https://github.com/jschicht/EaTools) mentioned EA is for OS/2 compatibility. So question is make sense to consider the API in modern world?

**Comment by heinrich-ulbricht** [comment]
@iSazonov This article shows active development regarding EAs: [User Xattr Support Finally Landing For NFS In Linux 5.9](https://www.phoronix.com/scan.php?page=news_item&px=Linux-5.9-NFS-Server-User-Xattr):

> The NFS server updates for Linux 5.9 have support for user-extended attributes on NFS. This is the functionality outlined via IETF's RFC 8276 for handling of file-system extended attributes in NFSv4

Now that there seem to be two options to implement EA-compatibility on Windows - EAs and ADS - there are different goals one might try to achieve:

1. provide an API to access ADS on Windows - not the goal of this specific ticket
2. provide an API to access EAs on any platform - although the ticket title says this is the goal it now feels too technical as I don't care which underlying technology will be used to store user-controller metadata
3. provide a means to store user-controlled file metadata on any one platform - we could achieve this by providing an API that uses ADS on Windows and EAs on Linux et al., but metadata wouldn't be preserved when moving files across file systems; so this is not enough
4. provide a means to store user-controlled file metadata in a cross-platform and cross-file system way - **THIS** is the goal of this ticket

My need when opening this ticket was to put user-controlled metadata on files that is persisted when moving those files across file system boundaries ([ext2, ext3, ext4, JFS, Squashfs, Yaffs2, ReiserFS, Reiser4, XFS, Btrfs, OrangeFS, Lustre, OCFS2 1.6, ZFS, and F2FS](https://en.wikipedia.org/wiki/Extended_file_attributes#Linux) - and NTFS). And to have a .NET API to set this metadata in a cross-platform .NET application.

**Comment by heinrich-ulbricht** [comment]
Oh and I don't know if somebody tried to be funny here. Citation from [IETF's RFC 8276](https://tools.ietf.org/html/rfc8276):

> This document describes an optional feature extending the NFSv4
>    protocol.  This feature allows extended attributes (hereinafter also
>    referred to as xattrs) to be interrogated and manipulated using NFSv4
>    clients.  Xattrs are provided by a file system to associate opaque
>    metadata, not interpreted by the file system, with files and
>    directories.  _**Such support is present in many modern local file
>    systems.**_

(Emphasis mine.)

**Comment by mklement0** [comment]
Thanks for the sleuthing, @iSazonov.

So it does sound like EAs and ADS are alternative implementations of the same basic functionality - but for compatibility with _different_ operating systems, given that the blog post that @heinrich-ulbricht [linked to](https://blogs.msmvps.com/alunj/2020/06/24/revisiting-ntfs-alternate-data-streams/) states: 

> Alternate Data Streams were originally created to support Apple Mac Resource Forks, in files copied from Apple to NTFS and back. I’m not sure Apple even bothers with them any more, now that they’ve moved to something akin to Linux as their OS. [Indeed]

The [EA tools repo](https://github.com/jschicht/EaTools) you mention additionally states the following, disconcerting fact (emphasis added):

> **Windows does not contain any API that can be used to _remove_ extended attributes**. That means if we want to remove $EA from a file, we can:
> * Delete the file altogether.
> * Move the file to a non-NTFS volume and back again.
> * Archive the file to a zip (for instance), delete the original file, and then unpack the zip back to the original location.
> * Modify $MFT directly. This may work in certain circumstances.

Anecdotally, I can say that while I've come across ADS many times, I had never heard of EAs before - which, given the above, is perhaps not too surprising.

Also, the NFSv4-related [RFC  8276](https://github.com/dotnet/runtime/issues/49604#issuecomment-833263735) @heinrich-ulbricht links to with respect to NFS references ADS (though not by that exact name):

> In the New Technology File System (NTFS), extended attributes may be stored within "file streams" [[NTFS]](https://docs.microsoft.com/en-us/windows/win32/fileio/file-streams)

**Comment by iSazonov** [comment]
> **Windows does not contain any API that can be used to _remove_ extended attributes**.

Perhaps ZwSetEaFile, can do this.

**Comment by KalleOlaviNiemitalo** [comment]
Because named streams on Windows can already be created, read, and deleted by passing the appropriate strings to System.IO.FileStream and such, I think there is less need for an alternative "extended file attributes" API that does the same thing; that's why it might be more useful to have the API access EAs as implemented in NTFS. It appears .NET doesn't currently have an API for _enumerating_ named streams ([FindFirstStreamW](https://docs.microsoft.com/windows/win32/api/fileapi/nf-fileapi-findfirststreamw)), though.

**Comment by iSazonov** [comment]
> It appears .NET doesn't currently have an API for _enumerating_ named streams ([FindFirstStreamW](https://docs.microsoft.com/windows/win32/api/fileapi/nf-fileapi-findfirststreamw)), though.

Yes, and PowerShell implements this internally (GetStreams() method) so I think it makes sense to consider the API in .Net.

**Comment by heinrich-ulbricht** [comment]
@mklement0 
> I do not think this is feasible, as the only feature _guaranteed_ to be available across file-systems is file _content_, so you'd have to store the metadata as part of the content, which in turn means that you require special tools / APIs to read the regular content - such a file won't act like a regular file anymore (and for _directories_ you'd be out of luck anyway).

Hm, modifying file content does not seem like an option for arbitrary files if the application doesn't "own" the files. And you are right, there is no guarantee the EAs survive when files (or directories) are moved between file systems. But I think from an API perspective I expect to be able to CRUD EAs in .NET on different platforms. This at least would make applications possible that use EAs to store application-specific metadata. Whether EAs survive moving to another file system is out of scope for the API.

Reading the comments I see the hesitancy to pull this "old" EA feature out of the closet into a modern API. It would be good to know how widespread the use of EAs currently really is - or estimate how widespread it would be if there was an API.

Whatever the case I'm loving the discussion so far. Cross-platform, cross-file system, cross-team and cross-century. 🤩

**Comment by heinrich-ulbricht** [comment]
Another thought: if it **won't** be possible to provide either

- a .NET API to access native EAs, or
- a .NET adapter API to transparently access either EAs (non-NTFS) or ADS (NTFS)

then a .NET application would have to 

- check for the availability of ADS (how?), if yes use them, if no
- use companion-binaries that are platform/file system specific and allow accessing EAs - those could be cmdline tools from the ancient past, it doesn't matter, as long as they can access EAs

The latter sounds like a new project to be had that provides an adapter API for those tools so they could be used via one API on any platform. This would be the not-so-clean but pragmatic approach.

**Comment by heinrich-ulbricht** [comment]
Just adding another data point for the topic of EAs on Windows: [Appropriate function to set NTFS extended attributes: ZwSetEaFile or NtSetEaFile](https://reverseengineering.stackexchange.com/questions/17310/appropriate-function-to-set-ntfs-extended-attributes-zwseteafile-or-ntseteafile)

**Comment by heinrich-ulbricht** [comment]
Found more info about xattr and cross-platform support over here in this discussion in the borg backup repo: https://github.com/borgbackup/borg/issues/1342#issuecomment-233146128

And again, a rather amusing take on this topic: 

> Alternate Data Streams (ADS) are the Windows equivalent of resource forks (and/or extended attributes, **depending on platform and the moon-phase when you're asking the question ;)**

Also some code that respects different platforms: https://github.com/borgbackup/borg/blob/master/src/borg/xattr.py

**Comment by EraYaN** [comment]
Especially on Linux this is very useful to set (and remove) SELinux attributes or for example the NTACL/DOSATTRIB xattrs that samba leaves behind. So the underlying technology does matter (for that use case at least), it shouldn't just be a way to store random metadata in "some" store. The exact keys/names and tech are important so other native tools can use the same attributes. 

Right now you can call the actual setxattr functions directly or use the Mono.Posix package, but it's not as nice.

**Comment by ProIntegritate** [comment]
It would be good to have access to this functionality, just setting and reading from ADS/EA would be sufficient to me, also as a byte array as not all data is in English and ASCII strings. I don't care about deleting ADS/EA. Storing extra meta information about files is kind of useful, as well as being able to scan files for alternative data streams without messing around with Win32 API and PInvoke crap, something i'd prefer not to do anymore. 

In fact it would be great if .NET would grow and import legacy API stuff, regardless of cross platform support. Cross platform functionality can be added in time - if it exists.

**Comment by heinrich-ulbricht** [comment]
Note: came across this nice sample of creating and copying EAs, leaving it here, in case somebody wants to see that in action: https://github.com/gtworek/PSBits/tree/master/CopyEAs

**Comment by heinrich-ulbricht** [comment]
Another sample that reads EAs on Windows: https://gist.github.com/jborean93/50a517a8105338b28256ff0ea27ab2c8

Today have another use case for Alternate Data Streams that needs to work cross-platform. And so this topic came back :D.
